zarr/core.py

```diff
@@ -1769,7 +1775,7 @@ def _encode_chunk(self, chunk):
                 chunk = f.encode(chunk)
 
         # check object encoding
-        if isinstance(chunk, np.ndarray) and chunk.dtype == object:
+        if ensure_ndarray(chunk).dtype == object:
```

Comment:
Is it possible that some non-`ndarray`s would previously skip over this `if` block but now fail if `ensure_ndarray` raises an error? For example:

```python
In [28]: from numcodecs.compat import ensure_ndarray

In [29]: import array

In [30]: chunk = array.array('u', 'hello \u2641')

In [31]: if isinstance(chunk, np.ndarray) and chunk.dtype == object:
    ...:     pass
    ...:

In [32]: if ensure_ndarray(chunk).dtype == object:
    ...:     pass
    ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-e1e15cbf81c2> in <module>
----> 1 if ensure_ndarray(chunk).dtype == object:
      2     pass
      3

~/miniconda/envs/zarr-python-dev/lib/python3.7/site-packages/numcodecs/compat.py in ensure_ndarray(buf)
     63         # problematic because numpy does not support it on all platforms. Also do not
     64         # support char as it was removed in Python 3.
---> 65         raise TypeError('array.array with char or unicode type is not supported')
     66
     67     else:

TypeError: array.array with char or unicode type is not supported
```

To be clear, I'm not sure how likely this is (or if it's even possible) to come up in practice. Do you have a sense for this?

Reply:
That exception would be expected as we have decided not to work with Python builtin arrays that use character or unicode types. There is some more detailed discussion in [this thread]( https://github.com/zarr-developers/numcodecs/pull/128#discussion_r236476705 ).

Reply:
Basically the only exception that we need to worry about is [the one coming from `memoryview`]( https://github.com/zarr-developers/numcodecs/blob/8de24ba51492c08bfe6898e209a0a05727f928bc/numcodecs/compat.py#L68 ), which will be a `TypeError` if the object doesn't support [the buffer protocol]( https://docs.python.org/3/c-api/buffer.html ). After that point we have a `memoryview`. So we know the rest of the code will work (unless NumPy gets a bug ;).

Reply:
For context this function is trying to turn some data into bytes that can be serialized (say to a file on disk).

The filters before this step should be returning something that either is an `ndarray` or could be coerced to one. This is required by the compressors that follow and the storage layer afterwards. So if `ensure_ndarray` fails here, then we have invalid data and raising an exception to the user would be appropriate.